edx_prefectutils/amplitude.py

```diff
@@ -0,0 +1,347 @@
+            last_idx_for_worker = len(accumulated_events)
+        else:
+            if accumulated_events[0].user_device == accumulated_events[batch_size - 1].user_device:
+                # We have a segment which is larger than a single bucket.  We should find the last event for that
+                # segment and send the entire segment to a worker.
+                target_user_device = accumulated_events[0].user_device
+                idx = batch_size - 1
+                while True:
+                    # Is there a more efficient way than indexing one event at a time?  Yes, but simple iteration is not
+                    # our performance bottleneck, and anything more complicated can introduce bugs.
+                    idx += 1
+                    try:
+                        if accumulated_events[idx].user_device != target_user_device:
+                            last_idx_for_worker = idx
+                            break
+                    except IndexError:
+                        last_idx_for_worker = idx
+                        break
+            else:
+                # One or more segments completely fit into one batch, so lets consolidate them and send them.  Exclude
+                # the last segment so that we send no more than one-batch-worth of events by finding the first event for
+                # the last segment and using that as a cutoff:
+                cut_user_device = accumulated_events[batch_size - 1].user_device
+                idx = batch_size - 1
+                while True:
+                    idx -= 1
+                    if accumulated_events[idx].user_device != cut_user_device:
+                        last_idx_for_worker = idx + 1
+                        break
+
+        # Now we have `last_idx_for_worker` set to the point where we want to slice off events to send to a worker
+        # thread.  Also, it is important that we delete that slice from the accumulated_events so that the python
+        # garbage collector will free all that memory as soon as the corresponding worker thread returns.
+        events_for_next_worker = accumulated_events[:last_idx_for_worker]
+        del accumulated_events[:last_idx_for_worker]
+
+        # Finally, batch the events such that each batch is small enough to fit inside a single Amplitude API call.
+        yield batch_generator(events_for_next_worker, batch_size)
+
+
+def _get_old_high_watermark(sf_credentials: SFCredentials, sf_role: str):
+    """
+    Get the amplitude event ID of the event immediately following the very last event sent to Amplitude's API.
+
+    Raises:
+        snowflake.connector.ProgrammingError: if the loader log table is inaccessible for some reason.
+    """
+    sf_connection = create_snowflake_connection(sf_credentials, sf_role)
+    query = """
+        select max(amplitude_event_id)
+          from prod.amplitude_events.amplitude_events_loader_log
```

Comment:
At some point, we'll want to pull these out into constants or something -- AMPLITUDE_EVENT_DATABASE, AMPLITUDE_EVENT_SCHEMA, AMPLITUDE_EVENT_LOADER_LOG_TABLE -- stuff like that.   Not important now, though.   Though it might be useful in testing, when pointing elsewhere.